system/lib/pthread/proxying.c

```diff
@@ -580,3 +582,61 @@ em_promise_t emscripten_proxy_promise(em_proxying_queue* q,
                           &block->ctx,
                           &block->promise_ctx);
 }
+
+typedef struct proxied_js_func_t {
+  int funcIndex;
+  pthread_t callingThread;
+  int numArgs;
+  int64_t* argBuffer;
+  double result;
+} proxied_js_func_t;
+
+static void run_js_func(void* arg) {
+  proxied_js_func_t* f = (proxied_js_func_t*)arg;
+  f->result = _emscripten_receive_on_main_thread_js(
+    f->funcIndex, f->callingThread, f->numArgs, f->argBuffer);
+}
+
+static void run_js_done(void* arg) {
+  proxied_js_func_t* f = (proxied_js_func_t*)arg;
+  free(f);
+}
+
+static void run_js_cancelled(void* arg) {
+  proxied_js_func_t* f = (proxied_js_func_t*)arg;
+  assert(false);
```

Comment:
Would it be worth at least attaching a message to this assertion?

Reply:
I was wondering if I should maybe just not specify this at all.. WDTY?   If can I leave this callback as NULL and save on code size?   I think the in previous code there was no way to detect async failures so it would not be a regresion.

Reply:
Do we expect it to even be possible for the call to be canceled? That would happen if the main runtime thread shut down after the proxying message was sent but before it was received. If that is possible, I guess it's nice in theory to free the memory. Otherwise, or if it seems rare enough that we don't need to care about freeing the memory, it seems fine to have a NULL cancellation callback.

Reply:
Removed this code